addons/http_routing/models/ir_http.py

```diff
@@ -608,61 +598,31 @@ def _get_error_html(cls, env, code, values):
         return code, env['ir.ui.view']._render_template('http_routing.%s' % code, values)
 
     @classmethod
-    def _handle_exception(cls, exception):
-        is_frontend_request = bool(getattr(request, 'is_frontend', False))
-        if not is_frontend_request:
-            # Don't touch non frontend requests exception handling
-            return super(IrHttp, cls)._handle_exception(exception)
-        try:
-            response = super(IrHttp, cls)._handle_exception(exception)
-
-            if isinstance(response, Exception):
-                exception = response
-            else:
-                # if parent excplicitely returns a plain response, then we don't touch it
-                return response
-        except Exception as e:
-            if 'werkzeug' in config['dev_mode']:
-                raise e
-            exception = e
-
-        code, values = cls._get_exception_code_values(exception)
+    def _http_handle_error(cls, exception):
+        response = super()._http_handle_error(exception)
 
-        if code is None:
-            # Hand-crafted HTTPException likely coming from abort(),
-            # usually for a redirect response -> return it directly
-            return exception
+        is_frontend_request = bool(getattr(request, 'is_frontend', False))
+        if not is_frontend_request or not isinstance(response, HTTPException):
+            # neither handle backend requests nor plain responses
+            return response
 
-        if not request.uid:
+        if not request.env.uid:
+            # grant public user
             cls._auth_method_public()
 
-        # We rollback the current transaction before initializing a new
-        # cursor to avoid potential deadlocks.
+        if not hasattr(request, 'website'):
+            # setup website on the request
+            request.env['ir.http']._frontend_pre_dispatch()
```

Comment:
Isn't that weird to check website on request in http_routing?
Also, `# setup website on the request` is not true

Reply:
Why not do it all the time when `is_frontend_request` is true? And the condition `if not hasattr(request, 'website'):` in the inheritance of `_frontend_pre_dispatch` in the website?

Reply:
Actually `_frontend_pre_dispatch()` is idempotent so I dropped the condition

Reply:
Gorash,`_frontend_pre_dispatch()` uses the ORM thus we need a valid `user` to be set on the environment. We can only call `_frontend_pre_dispatch()` *after* `_authenticate`, this explains why it cannot be called before. 

The call to `_frontend_pre_dispatch` in:

1. `_pre_dispatch` covers the case when we matched a `website=True` route. (website specific)
2. `_serve_fallback` covers the case when a website page have been requested. When no backend route matched. (website specific)
3. `_http_handle_error` covers the case when there was any other error (including the 404 Not Found error when no backend route matched and that portal is installed without website). (portal generic)